src/block.rs

```diff
@@ -1,140 +1,191 @@
+impl<C: Codec, M: MultihashDigest> Block<C, M> {
+    /// Creates a new block.
+    pub fn new(cid: Cid, data: Box<[u8]>) -> Self {
+        Self {
+            _marker: PhantomData,
+            cid,
+            data,
+            vis: Visibility::Public,
+        }
     }
-    let hash = Box::<dyn MultihashDigest<H>>::from(cid.hash().algorithm()).digest(&data);
-    if hash.as_ref() != cid.hash() {
-        return Err(Error::InvalidHash(hash.to_vec()));
+
+    /// Returns the blocks visibility.
+    pub fn visibility(&self) -> Visibility {
+        self.vis
     }
-    raw_decode::<C, H, O, D>(cid.codec(), data)
-}
 
-/// Decode block to ipld.
-pub fn decode_ipld<C, H>(cid: &CidGeneric<C, H>, data: &[u8]) -> Result<Ipld<C, H>>
-where
-    C: Into<u64> + TryFrom<u64> + Copy + PartialEq + EncodeDecodeIpld<H>,
-    H: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    Box<dyn MultihashDigest<H>>: From<H>,
-{
-    if data.len() > MAX_BLOCK_SIZE {
-        return Err(Error::BlockTooLarge(data.len()));
+    /// Sets the blocks visibility.
+    pub fn set_visibility(&mut self, vis: Visibility) {
+        self.vis = vis;
     }
-    let hash = Box::<dyn MultihashDigest<H>>::from(cid.hash().algorithm()).digest(&data);
-    if hash.as_ref() != cid.hash() {
-        return Err(Error::InvalidHash(hash.to_vec()));
+
+    /// Encode a block.`
+    pub fn encode<T: Encode<C> + ?Sized>(codec: C, hcode: u64, payload: &T) -> Result<Self> {
+        let data = codec.encode(payload)?;
+        let cid = create_cid::<M>(codec.into(), hcode, &data)?;
+        Ok(Self {
+            _marker: PhantomData,
+            cid,
+            data,
+            vis: Visibility::Public,
+        })
+    }
+
+    /// Encode ipld.
+    pub fn encode_ipld(codec: C, hcode: u64, ipld: &Ipld) -> Result<Self> {
+        let data = codec.encode_ipld(ipld)?;
+        let cid = create_cid::<M>(codec.into(), hcode, &data)?;
+        Ok(Self {
+            _marker: PhantomData,
+            cid,
+            data,
+            vis: Visibility::Public,
+        })
+    }
+
+    /// Decodes a block.
+    pub fn decode<T: Decode<C>>(&self) -> Result<T> {
+        verify_cid::<M>(&self.cid, &self.data)?;
+        C::try_from(self.cid.codec())?.decode(&self.data)
     }
-    cid.codec()
-        .decode(data)
-        .map_err(|e| Error::CodecError(Box::new(e)))
-}
 
-/// Returns the references in an ipld block.
-pub fn references<C, H>(ipld: &Ipld<C, H>) -> HashSet<CidGeneric<C, H>>
-where
-    C: Into<u64> + TryFrom<u64> + Copy + Eq,
-    H: Into<u64> + TryFrom<u64> + Copy + Eq,
-{
-    let mut set: HashSet<CidGeneric<C, H>> = Default::default();
-    for ipld in ipld.iter() {
-        if let Ipld::Link(cid) = ipld {
-            set.insert(cid.to_owned());
+    /// Decodes to ipld.
+    pub fn decode_ipld(&self) -> Result<Ipld> {
+        verify_cid::<M>(&self.cid, &self.data)?;
+        C::try_from(self.cid.codec())?.decode_ipld(&self.data)
+    }
+
+    /// Convert from block.
```

Comment:
It would be nice to have a bit more explanation what to use it for and perhaps even an example.

Reply:
A Multicodec can only encode/decode to/from ipld for the mentioned reason. So if you want to use more "features" you need to create a codec specific block and convert it to a block that the store supports

Reply:
Double checking if I understand it correctly. I use a `Multicodec` based `Block` and then I want to store it in a storage that is bould to a `DagCbor` based `Block. Then I can use this method to transform it.

Reply:
If you get a block from a store you'll get a Multicodec/Multihash block. To decide it to a native types you need to convert it to a DagCbor/Multihash block. The cache is a good example